File: 733_flood_fill.py
class Solution(object):
    def floodFill(self, image, sr, sc, color):
        """
        :type image: List[List[int]]
        :type sr: int
        :type sc: int
        :type color: int
        :rtype: List[List[int]]
        """
        queue = [(sr, sc)]
        start_color = image[sr][sc]
        x_max = len(image) - 1
        y_max = len(image[0]) - 1

        if start_color == color:
            return image

        while len(queue) != 0:
            pixel = queue.pop()
            px = pixel[0]
            py = pixel[1]
            if image[px][py] == start_color:
                image[px][py] = color
                if px > 0 and image[px - 1][py] != color:
                    queue.append((px - 1, py))
                if px < x_max and image[px + 1][py] != color:
                    queue.append((px + 1, py))
                if py > 0 and image[px][py - 1] != color:
                    queue.append((px, py - 1))
                if py < y_max and image[px][py + 1] != color:
                    queue.append((px, py + 1))
        
        return image

File: c++/733_flood_fill.cpp
class Solution {
public:
    vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc, int color) {
        if (image[sr][sc] == color) {
            return image;
        }

        queue<tuple<int, int>> tofill;
        tofill.push(make_tuple(sr, sc));
        int org_color = image[sr][sc];
        int row_max = image.size() - 1, col_max = image[0].size() - 1;

        while (!tofill.empty()) {
            tuple<int, int> coords = tofill.front();
            int r = get<0>(coords), c = get<1>(coords);
            
            if (image[r][c] == org_color) {
                image[r][c] = color;

                if (r > 0 && image[r-1][c] == org_color) {
                    tofill.push(make_tuple(r-1, c));
                }
                if (r < row_max && image[r+1][c] == org_color) {
                    tofill.push(make_tuple(r+1, c));
                }
                if (c > 0 && image[r][c-1] == org_color) {
                    tofill.push(make_tuple(r, c-1));
                }
                if (c < col_max && image[r][c+1] == org_color) {
                    tofill.push(make_tuple(r, c+1));
                }
            } 

            tofill.pop();
        }

        return image;
    }
};
